File: modules/mrz-scanner/ios/MRZScannerModule.swift
import ExpoModulesCore
import UIKit

public class MRZScannerModule: Module {
    public func definition() -> ModuleDefinition {
        Name("MRZScanner")

        AsyncFunction("scanMRZ") { (options: [String: Any]?, promise: Promise) in
            let instructionText = (options?["instructionText"] as? String)
            let isChipShow = (options?["isChipShow"] as? Bool) ?? true
            let timeoutMs = (options?["timeoutMs"] as? Double) ?? 0
            DispatchQueue.main.async {
                self.presentScanner(promise: promise, instructionText: instructionText, isChipShow: isChipShow, timeoutMs: timeoutMs)
            }
        }
    }

    private func presentScanner(promise: Promise, instructionText: String?, isChipShow: Bool = true, timeoutMs: Double = 0) {
        guard let presenter = topViewController() else {
            // Fix #13: English error message (was Turkish)
            promise.reject("ERR_UI", "Scanner could not be opened: no active view controller found.")
            return
        }

        var completed = false

        let scanner = VisionViewController()
        scanner.modalPresentationStyle = .fullScreen

        if let text = instructionText, !text.isEmpty {
            scanner.instructionText = text
        }

        scanner.isChipShow = isChipShow

        scanner.completionHandler = { mrz in
            guard !completed else { return }
            completed = true
            scanner.dismiss(animated: true) {
                promise.resolve(mrz)
            }
        }

        scanner.onCancel = {
            guard !completed else { return }
            completed = true
            // Fix #7: English error message
            promise.reject("ERR_CANCELLED", "MRZ scan was cancelled.")
        }

        presenter.present(scanner, animated: true)

        // Fix #10: timeout support
        if timeoutMs > 0 {
            DispatchQueue.main.asyncAfter(deadline: .now() + timeoutMs / 1000.0) {
                guard !completed else { return }
                completed = true
                scanner.dismiss(animated: true) {
                    promise.reject("ERR_TIMEOUT", "MRZ scan timed out after \(Int(timeoutMs))ms.")
                }
            }
        }
    }

    private func topViewController(base: UIViewController? = nil) -> UIViewController? {
        let baseController: UIViewController?

        if let base {
            baseController = base
        } else {
            let scenes = UIApplication.shared.connectedScenes
                .compactMap { $0 as? UIWindowScene }
                .filter { $0.activationState == .foregroundActive }
            let window = scenes
                .flatMap { $0.windows }
                .first { $0.isKeyWindow }
            baseController = window?.rootViewController
        }

        if let nav = baseController as? UINavigationController {
            return topViewController(base: nav.visibleViewController)
        }

        if let tab = baseController as? UITabBarController {
            return topViewController(base: tab.selectedViewController)
        }

        if let presented = baseController?.presentedViewController {
            return topViewController(base: presented)
        }

        return baseController
    }
}

File: modules/mrz-scanner/android/src/main/java/cc/getportal/mrz/MRZScannerModule.kt
package cc.getportal.mrz

import android.app.Activity
import android.content.Intent
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

class MRZScannerModule : Module() {
    companion object {
        private const val REQUEST_CODE_SCAN = 9101
        const val EXTRA_MRZ_RESULT = "mrz_result"
        const val EXTRA_ERROR = "mrz_error"
    }

    private var pendingPromise: Promise? = null

    override fun definition() = ModuleDefinition {
        Name("MRZScanner")

        AsyncFunction("scanMRZ") { options: Map<String, Any>?, promise: Promise ->
            val activity = appContext.activityProvider?.currentActivity
            if (activity == null) {
                // Fix #7: English error message (was Turkish)
                promise.reject("ERR_NO_ACTIVITY", "No activity found.", null)
                return@AsyncFunction
            }

            // Fix #5: reject orphaned promise before overwriting
            pendingPromise?.reject("ERR_CANCELLED", "A new scan was started.", null)
            pendingPromise = promise

            val intent = Intent(activity, MRZScannerActivity::class.java)
            val instructionText = options?.get("instructionText") as? String
            if (!instructionText.isNullOrEmpty()) {
                intent.putExtra(MRZScannerActivity.EXTRA_INSTRUCTION_TEXT, instructionText)
            }
            val isChipShow = (options?.get("isChipShow") as? Boolean) ?: true
            intent.putExtra(MRZScannerActivity.EXTRA_IS_CHIP_SHOW, isChipShow)

            // Fix #10: pass timeout to activity
            val timeoutMs = (options?.get("timeoutMs") as? Number)?.toLong() ?: 0L
            intent.putExtra(MRZScannerActivity.EXTRA_TIMEOUT_MS, timeoutMs)

            activity.startActivityForResult(intent, REQUEST_CODE_SCAN)
        }

        OnActivityResult { _, payload ->
            val promise = pendingPromise ?: return@OnActivityResult
            pendingPromise = null

            val resultCode = payload.resultCode
            val data = payload.data

            if (resultCode == Activity.RESULT_OK && data != null) {
                val mrz = data.getStringExtra(EXTRA_MRZ_RESULT)
                if (!mrz.isNullOrEmpty()) {
                    promise.resolve(mrz)
                } else {
                    // Fix #7: English error message
                    promise.reject("ERR_MRZ", "MRZ could not be read.", null)
                }
            } else {
                val error = data?.getStringExtra(EXTRA_ERROR)
                promise.reject("ERR_CANCELLED", error ?: "MRZ scan was cancelled.", null)
            }
        }
    }
}
